OpenChemistry/avogadrolibs
# chore: update includes with include-what-you-use

Hacktoberfest 2023

This PR updates the avogadrolibs submodule `includes` using the include-what-you-use tool.

Note: there were some build issues after running the tool; specifically with some missed includes on `<fstream>` and '<sstream>`, so these were manually added back in where needed.

#### The Process

The process was non-trivial and not particularly fun.  I'm including an outline of the steps here for anyone in the future wishing to do the same.

1. Fork the main `openchemistry` project
2. Fork the specific submodule you will be working with (in this case `avogadrolibs`)
3. Clone the main `openchemistry` repo to your machine
4. Edit the submodule file to point the submodule in question (`avogadrolibs`) to your fork of the submodule
5. Clone the submodule repo

Once you have all the code on your machine.  Time to set-up the toolchain.

1. You will need the correct version of CMake
2. You will need the latest version of include-what-you-use
3. You will need the correct version of Python
4. You will need the correct version of gcc and/or g++ (or whatever compiler you will use).

Now that you have your toolchain built, it's time to go through the process.

1. Run CMake to configure the build
2. Run CMake to build the project
3. Once you have a working build, it's time to run with include-what-you-use
4. Edit the submodule's CMakeFiles.txt file to include the following lines:
> `line 1`
> `line2`
5. Run CMake build again, but this time redirect the output to a file
`EXAMPLE:`
6. Now you can run the iwyu_tool.py script, directing the output file as input to the script.  This will perform the code updates.  Note: this can take a while.  Like a really long time.  Like.  REALLY long.  And there's no progress indicator.  It may seem like the process is frozen.  Just wait.  Go get a snack.  Watch some tv.
` $ iwyu_tool.py < iwyu_ouput.txt`
7. Once this is done, better check that it worked.  Revert your changes to the CMakeFiles.txt file.  Then build the project with CMake again.
8. Fix any errors.
9. Repeat steps 4 through 8 until everything builds clean.
10. commit your code
11. push to your repo
12. realize you forgot to run this in your fork of the submodule.  GOTO: 1


<hr>

Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
